sharktank/sharktank/layers/paged_attention.py

```diff
@@ -315,116 +331,201 @@ def write_timestep(
         devices = self.pipeline_to_device_map[pipeline]
         transformer_block_count = self.pipeline_to_block_count[pipeline]
         block_offset = self.pipeline_to_block_offset[pipeline]
+        transformer_block_index = transformer_block_index - block_offset
 
+        page_tables = self.unflatten_page_tables(state)
         page_table = page_tables[pipeline]
         page_table = page_table.flatten(0, 3)
-        bs, *_ = seq_positions.shape
         assert len(cache_partitions) == self.cache_partition_count
 
-        transformer_block_index = transformer_block_index - block_offset
+        for p, partition in enumerate(cache_partitions):
+            if not isinstance(page_table, ShardedTensor):
+                self.write_timestep_shard(
+                    page_table=page_table,
+                    partition_id=p,
+                    cache_partition=partition,
+                    transformer_block_index=transformer_block_index,
+                    transformer_block_count=transformer_block_count,
+                    seq_positions=seq_positions,
+                    page_ids=page_ids,
+                )
+                continue
+
+            if (
+                isinstance(page_table, ShardedTensor)
+                and isinstance(partition, ShardedTensor)
+                and isinstance(seq_positions, ReplicatedTensor)
+                and isinstance(page_ids, ReplicatedTensor)
+            ):
+
+                page_table_split = isinstance(page_table, SplitPrimitiveTensor)
+                partition_split = isinstance(partition, SplitPrimitiveTensor)
+                assert page_table_split == partition_split
+
+                page_table_shards = page_table.shards
+                partition_shards = self.deshard_like(page_table, partition)
+                seq_shards = self.deshard_like(page_table, seq_positions)
+                page_ids_shards = self.deshard_like(page_table, page_ids)
+
+                for i in range(page_table.shard_count):
+                    self.write_timestep_shard(
+                        page_table=page_table_shards[i],
+                        partition_id=p,
+                        cache_partition=partition_shards[i],
+                        transformer_block_index=transformer_block_index,
+                        transformer_block_count=transformer_block_count,
+                        seq_positions=seq_shards[i],
+                        page_ids=page_ids_shards[i],
+                    )
+                continue
 
-        # [bs, 1, atten_head_count, attn_head_dim]
-        for idx, cache_partition in enumerate(cache_partitions):
-            # [bs, 1]
-            page_index = seq_positions // self.block_seq_stride
+            assert False
 
-            page_id = ops.gather(page_ids, dim=1, index=page_index.unsqueeze(1))
-            page_offset = (seq_positions % self.block_seq_stride).unsqueeze(1)
+    def write_timestep_shard(
```

Comment:
I'm not really sure what is happening here. There are a lot of sharding details but no documentation on what we are doing. I'm guessing you are splitting things into sharded and unsharded versions for each method. Can you document the sharding details and what parts are the unsharded versions? This is very hard to follow.